### PR TITLE
GraphQL and TS for fetching and editing buckets

### PR DESCRIPTION
**Pull request by nl0**
# Description

Extracted from #2212 

Use graphql for:
- containers/Admin/Buckets
- containers/Bucket/Overview
- utils/BucketConfig
- utils/LinkedData (not fully functional yet, so i disabled it for now)

# TODO

- [x] [Changelog](CHANGELOG.md) entry

### CODE REVIEW

File: catalog/app/containers/Bucket/PackageTree.js [hunk 1]
```diff
@@ -611,13 +618,15 @@ function PackageTree({ bucket, mode, name, revision, path, resolvedFrom }) {
   return (
     <FileView.Root>
       <MetaTitle>{[`${name}@${R.take(10, revision)}/${path}`, bucket]}</MetaTitle>
+      {/* TODO: bring back linked data after re-implementing it using graphql
```
Comment (fiskus) [inline catalog/app/containers/Bucket/PackageTree.js:621]: Maybe, you wanted to re-implement it in this PR and forgot?

File: catalog/app/utils/BucketConfig.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import * as urql from 'urql'
+
+import * as AuthSelectors from 'containers/Auth/selectors'
+import * as Config from 'utils/Config'
+import * as NamedRoutes from 'utils/NamedRoutes'
+import { useRoute } from 'utils/router'
+
+import BUCKET_CONFIGS_QUERY from './BucketConfigList.generated'
+
+// always suspended
```
Comment (fiskus) [inline catalog/app/utils/BucketConfig.ts:13]: If this comment is for me too, I don't understand it

**Review by fiskus — APPROVED** [review]

File: catalog/app/containers/Admin/BucketConfigSelection.generated.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable @typescript-eslint/naming-convention */
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Admin/BucketConfigSelection.generated.ts:1]: correct

> The source of truth for these generated types somewhere in GraphQL?

for queries / mutations `path/to/file.graphql` becomes `path/to/file.generated.ts`, for schema it's a little different -- the schema file lives in the `shared` folder and generated code for it lives in `app/model/graphql`

File: catalog/app/containers/Admin/Buckets.tsx [hunk 4]
```diff
@@ -1,39 +1,157 @@
...
+}
+
+const SNS_ARN_RE = /^arn:aws(-|\w)*:sns:(-|\w)*:\d*:\S+$/
 
 const DO_NOT_SUBSCRIBE_STR = 'DO_NOT_SUBSCRIBE'
 const DO_NOT_SUBSCRIBE_SYM = Symbol(DO_NOT_SUBSCRIBE_STR)
 
-const SNS_ARN_RE = /^arn:aws(-|\w)*:sns:(-|\w)*:\d*:\S+$/
+type SnsFormValue = string | typeof DO_NOT_SUBSCRIBE_SYM
+// eslint-disable-next-line @typescript-eslint/no-redeclare
```
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Admin/Buckets.tsx:45]: it's a little tricky when you want to declare some kind of [type representative](https://github.com/fantasyland/fantasy-land#type-representatives) (couldnt find a better explanation for this term). like, when you want to define a `io-ts` codec with the same name as the type it represents for convenience, to avoid extra names like `SmthCodec` or `SmthValue` and such.
also, i see eslint rules as something that prevents us from doing some stupid things by accident, not as dogmas we must obey even if they stand in our way.

File: catalog/app/utils/BucketConfig.ts (hunk 2 above)
Comment (nl0) [reply catalog/app/utils/BucketConfig.ts:13]: previously, this hook had a `suspend` parameter, which allowed to opt-out of using suspense by setting it to false, but since its creation this feature hasnt been used, so i removed this option and made it always use suspense

File: catalog/app/containers/Bucket/PackageTree.js (hunk 1 above)
Comment (nl0) [reply catalog/app/containers/Bucket/PackageTree.js:621]: i started to migrate that code, but had not enough time to complete it, so it's still a TODO

File: catalog/app/containers/Admin/BucketConfigSelection.generated.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Admin/BucketConfigSelection.generated.ts:1]: but you should read a couple of them anyways to understand what code (types and data) gets generated from graphql documents

**Comment by sir-sigurd** [comment]
> @sir-sigurd pylint broke -- is it because of a new release or smth?

@nl0 
yup, fixed in https://github.com/quiltdata/quilt/commit/3c56e4fc0e8933201b54b660af3afef24736de72

File: catalog/app/containers/Admin/BucketConfigSelection.generated.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sir-sigurd) [reply catalog/app/containers/Admin/BucketConfigSelection.generated.ts:1]: @nl0 
if these files are generated automatically do we really need to put them in repo?

[the remark this answers is not in the document]
Comment (nl0) [reply catalog/app/containers/Admin/BucketConfigSelection.generated.ts:1]: well, that's a good question and i don't have a good answer.
TLDR: probably we do for now, but ideally we shouldnt.

i hoped that there exists some tool that would do just that -- generate ts code from graphql queries without writing the actual files to disc (keeping them virtual), but unfortunately typescript doesnt make it easy to load custom file types and transform them into ts in memory (there is a candidate for this -- https://github.com/firede/ts-transform-graphql-tag, but it's abandoned and doesnt work with the current versions of the libraries).
we could use a webpack loader, but ts autocomplete wont work with this approach.
also, we could generate the files on the fly every time development server is started (or as a build step in case of production builds), which may be ok, actually.
anyways, i plan to iterate on all this code generation logic soonish to reduce the maintenance burden as much as i can.

[the remark this answers is not in the document]
Comment (sir-sigurd) [reply catalog/app/containers/Admin/BucketConfigSelection.generated.ts:1]: > also, we could generate the files on the fly every time development server is started (or as a build step in case of production builds), which may be ok, actually.

That sounds reasonable.

> anyways, i plan to iterate on all this code generation logic soonish to reduce the maintenance burden as much as i can.

:+1:
